File: BYTES.NET/MEF/ExtensionsManager.vb
'import .net namespace(s) required
Imports System.ComponentModel.Composition.Hosting
Imports System.ComponentModel.Composition

Imports System.Collections.ObjectModel

Imports System.IO

Namespace MEF

    ''' <summary>
    ''' the basic MEF-based extensions manager
    ''' </summary>
    ''' <typeparam name="TInterface"></typeparam>
    ''' <typeparam name="TMetadata"></typeparam>
    Public Class ExtensionsManager(Of TInterface, TMetadata)

#Region "private variable(s)"

        <ImportMany()>
        Private _allExtensions As ObservableCollection(Of Lazy(Of TInterface, TMetadata)) = New ObservableCollection(Of Lazy(Of TInterface, TMetadata))
        Private _validExtensions As ObservableCollection(Of Lazy(Of TInterface, TMetadata)) = Nothing

#End Region

#Region "public properties"

        Public ReadOnly Property Extensions As Lazy(Of TInterface, TMetadata)()
            Get

                Return _validExtensions.ToArray

            End Get
        End Property

#End Region

#Region "public new instance method(s)"

        ''' <summary>
        ''' default new instance method(s)
        ''' </summary>
        Public Sub New()
        End Sub

#End Region

#Region "public method(s)"

        ''' <summary>
        ''' method for loading the extensions from path(s) given
        ''' </summary>
        ''' <param name="paths"></param>
        Public Sub Update(ByVal paths As String())

            'assemble the catalog
            Dim catalog = New AggregateCatalog()

            For Each path As String In paths

                'parse the path
                Dim tmp As Uri = Nothing
                If Uri.TryCreate(path, UriKind.Absolute, tmp) Then 'URIs will be kept and file/ folder paths will be converted to URI

                    path = tmp.LocalPath 'set the local path of the URI item

                End If

                'add extensions for a single file
                If File.Exists(path) Then

                    catalog.Catalogs.Add(New AssemblyCatalog(path))

                End If

                'add extensions for entire directories
                If path.Contains("*") Then

                    Dim pathSplit As String() = path.Split("\")
                    Dim lastElement As String = pathSplit(pathSplit.Length - 1)

                    If lastElement.Contains("*") Then 'the last element contains a wildcard

                        Dim rootPath As String = path.Substring(0, path.Length - (lastElement.Length + 1))

                        If Directory.Exists(rootPath) Then 'check for the root path

                            Dim rootDir As DirectoryInfo = New DirectoryInfo(rootPath)

                            If lastElement.Contains(".") Then 'the last element is a file

                                For Each file As FileInfo In rootDir.GetFiles(lastElement)

                                    catalog.Catalogs.Add(New AssemblyCatalog(file.FullName))

                                Next

                            Else 'the last element is a folder

                                For Each dir As DirectoryInfo In rootDir.GetDirectories(lastElement)

                                    catalog.Catalogs.Add(New DirectoryCatalog(dir.FullName))

                                Next

                            End If

                        End If

                    End If

                Else 'the last element contains no wildcard

                    If Directory.Exists(path) Then

                        catalog.Catalogs.Add(New DirectoryCatalog(path))

                    End If

                End If

            Next

            'load extensions from current assembly
            'catalog.Catalogs.Add(New AssemblyCatalog(System.Reflection.Assembly.GetExecutingAssembly()))

            'compose the collection
            Dim container As CompositionContainer = New CompositionContainer(catalog)
            container.ComposeParts(Me)

            'validate the extensions
            ValidateExtensions()

        End Sub

#End Region

#Region "private method(s)"

        ''' <summary>
        ''' method validating the extension(s) by type
        ''' </summary>
        Private Sub ValidateExtensions()

            _validExtensions = New ObservableCollection(Of Lazy(Of TInterface, TMetadata))

            For Each extension In _allExtensions

                Try

                    If ImplementsInterface(extension) Then

                        _validExtensions.Add(extension)

                    End If

                Catch ex As Exception

                End Try

            Next

        End Sub

        ''' <summary>
        ''' method checking if a class implements the 'TInterface' interface
        ''' </summary>
        ''' <param name="extension"></param>
        ''' <returns></returns>
        Private Function ImplementsInterface(ByRef extension) As Boolean

            For Each iFace In extension.Value.GetType.GetInterfaces

                If iFace.ToString = GetType(TInterface).ToString Then

                    Return True

                End If

            Next

            Return False

        End Function

#End Region

    End Class

End Namespace

File: src-2.x/BYTES.NET/IO/Extensibility/ExtensionsManager.cs
//import .net namespace(s) required
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

//import intenal namespace(s) required
using BYTES.NET.IO.Logging;
using BYTES.NET.IO.Extensibility.API;

namespace BYTES.NET.IO.Extensibility
{
    /// <summary>
    /// the (basic) extensions manager class
    /// </summary>
    /// <typeparam name="TInterface"></typeparam>
    public class ExtensionsManager<TInterface>
    {
        #region protected variable(s)

        protected Log _log = new Log() { CacheLimit = -1, Threshold = LogEntry.InformationLevel.Debug };
        protected Extension<TInterface>[] _extensions = new Extension<TInterface>[] { };

        protected Dictionary<string, string> _variables = new Dictionary<string, string>();
        protected IExtensionsSource[] _sources = new IExtensionsSource[] { };

        #endregion

        #region public properties

        public Extension<TInterface>[] Extensions
        {
            get => _extensions;
        }

        public Log Log
        {
            get => _log;
        }

        public Dictionary<string, string> Variables
        {
            get => _variables;
            set => _variables = value;
        }

        public IExtensionsSource[] Sources
        {
            get => _sources;
        }

        #endregion

        #region public new instance method(s)

        /// <summary>
        /// default new instance method
        /// </summary>
        /// <param name="paths"></param>
        public ExtensionsManager(string[] paths = null)
        {
            if (paths != null)
            {

                List<IExtensionsSource> sources = new List<IExtensionsSource>();
                foreach (string path in paths)
                {
                    sources.Add(new ExtensionsSource() { Source = path });
                }

                Update(sources.ToArray());

            }
        }

        /// <summary>
        /// overloaded new instance method
        /// </summary>
        /// <param name="sources"></param>
        public ExtensionsManager(IExtensionsSource[] sources)
        {
            Update(sources);
        }

        #endregion

        #region  public method(s)

        /// <summary>
        /// (re-)loads the extension(s) (from 'IExtensionsSource' list)
        /// </summary>
        /// <param name="sources"></param>
        /// <param name="resetLog"></param>
        public void Update(IExtensionsSource[] sources = null, bool resetLog = true)
        {

            //parse the argument(s)
            if (sources == null)
            {
                sources = _sources;
            }
            else
            {
                _sources = sources;
            }

            //reset the log (if requested)
            if (resetLog)
            {
                _log = new Log() { CacheLimit = -1 };
            }

            //get the extension(s)
            List<Extension<TInterface>> extensions = new List<Extension<TInterface>>();

            foreach (IExtensionsSource source in sources) //loop for each source
            {

                Lazy<TInterface>[] foundExtensions = source.Extensions<TInterface>(_variables);
                _log.Trace(foundExtensions.Length.ToString() + " extensions found from source '" + source.Source + "' using pattern '" + source.Pattern + "'");

                foreach (Extension<TInterface> extension in foundExtensions) //loop for each extension found
                {
                    extensions.Add(extension);
                }
            }

            _extensions = extensions.ToArray();
            _log.Trace(_extensions.Length.ToString() + " extensions found from " + sources.Length.ToString() + " sources");

        }

        /// <summary>
        /// overloaded method, (re-)loads the extensions (from string paths list)
        /// </summary>
        /// <param name="paths"></param>
        /// <param name="resetLog"></param>
        public void Update(string[] paths, bool resetLog = true)
        {

            List<IExtensionsSource> sources = new List<IExtensionsSource>();
            foreach (string path in paths)
            {
                sources.Add(new ExtensionsSource() { Source = path });
            }

            Update(sources.ToArray(), resetLog);

        }

        #endregion
    }

    /// <summary>
    /// the (more advanced) extensions manager
    /// </summary>
    /// <typeparam name="TInterface"></typeparam>
    /// <typeparam name="TMetadata"></typeparam>
    public class ExtensionsManager<TInterface, TMetadata> : ExtensionsManager<TInterface>
    {

        #region protected variable(s)

        new protected Extension<TInterface, TMetadata>[] _extensions = new Extension<TInterface, TMetadata>[] { };

        #endregion

        #region public properties

        new public Extension<TInterface, TMetadata>[] Extensions
        {
            get => _extensions;
        }

        #endregion

        #region public new instance method(s)

        /// <summary>
        /// default new instance method
        /// </summary>
        /// <param name="paths"></param>
        public ExtensionsManager(string[] paths = null) : base(paths)
        {
            if (paths != null)
            {
                Update(paths);
            }
        }

        /// <summary>
        /// overloaded new instance method
        /// </summary>
        /// <param name="sources"></param>
        public ExtensionsManager(IExtensionsSource[] sources) : base(sources)
        {
            Update(sources);
        }

        #endregion

        #region  public method(s)

        /// <summary>
        /// (re-)loads the extension(s) (from 'IExtensionsSource' list)
        /// </summary>
        /// <param name="sources"></param>
        /// <param name="resetLog"></param>
        new public void Update(IExtensionsSource[] sources = null, bool resetLog = true)
        {

            //parse the argument(s)
            if (sources == null)
            {
                sources = this._sources;
            }
            else
            {
                _sources = sources;
            }

            //reset the log (if requested)
            if (resetLog)
            {
                _log = new Log() { CacheLimit = -1 };
            }

            //get the extension(s)
            List<Extension<TInterface, TMetadata>> extensions = new List<Extension<TInterface, TMetadata>>();

            foreach (IExtensionsSource source in sources) //loop for each source
            {
                Lazy<TInterface, TMetadata>[] foundExtensions = source.Extensions<TInterface, TMetadata>(_variables);
                _log.Trace(foundExtensions.Length.ToString() + " extensions found from source '" + source.Source + "' using pattern '" + source.Pattern + "'");

                foreach (Extension<TInterface, TMetadata> extension in foundExtensions) //loop for each extension found
                {
                    extensions.Add(extension);
                }
            }

            _extensions = extensions.ToArray();
            _log.Trace(_extensions.Length.ToString() + " extensions found from " + sources.Length.ToString() + " sources");

        }

        /// <summary>
        /// overloaded method, (re-)loads the extensions (from string paths list)
        /// </summary>
        /// <param name="paths"></param>
        /// <param name="resetLog"></param>
        new public void Update(string[] paths, bool resetLog = true)
        {

            List<IExtensionsSource> sources = new List<IExtensionsSource>();
            foreach (string path in paths)
            {
                sources.Add(new ExtensionsSource() { Source = path });
            }

            Update(sources.ToArray(), resetLog);

        }

        #endregion
    }
}
